cmd/kubeadm/app/phases/upgrade/versiongetter.go

```diff
@@ -0,0 +1,139 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package upgrade
+
+import (
+	"fmt"
+	"io"
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	kubeadmutil "k8s.io/kubernetes/cmd/kubeadm/app/util"
+	versionutil "k8s.io/kubernetes/pkg/util/version"
+	"k8s.io/kubernetes/pkg/version"
+)
+
+// versionGetter defines an interface for fetching different versions.
+// Easy to implement a fake variant of this interface for unit testing
+type versionGetter interface {
+	GetClusterVersion() (string, *versionutil.Version, error)
+	GetKubeadmVersion() (string, *versionutil.Version, error)
+	GetLatestKubeVersion(string, string) (string, *versionutil.Version, error)
+	GetKubeletVersions() (map[string]int32, error)
+}
+
+// kubeVersionGetter handles the version-fetching mechanism from external sources
+type kubeVersionGetter struct {
+	client clientset.Interface
+	w      io.Writer
+}
+
+// Make sure kubeVersionGetter implements the versionGetter interface
+var _ versionGetter = &kubeVersionGetter{}
+
+// NewKubeVersionGetter returns a new instance of kubeVersionGetter
+func NewKubeVersionGetter(client clientset.Interface, writer io.Writer) *kubeVersionGetter {
+	return &kubeVersionGetter{
+		client: client,
+		w:      writer,
+	}
+}
+
+// GetClusterVersion gets API server version
+func (g *kubeVersionGetter) GetClusterVersion() (string, *versionutil.Version, error) {
+	fmt.Fprintf(g.w, "[upgrade/versions] Cluster version: ")
+	clusterVersionInfo, err := g.client.Discovery().ServerVersion()
+	if err != nil {
+		fmt.Fprintln(g.w, "<notfound>")
+		fmt.Fprintf(g.w, "[upgrade/versions] FATAL: Couldn't fetch cluster version from the API Server: %v\n", err)
+		return "", nil, err
+	}
+	fmt.Fprintln(g.w, clusterVersionInfo.String())
+
+	clusterVersion, err := versionutil.ParseSemantic(clusterVersionInfo.String())
+	if err != nil {
+		fmt.Fprintf(g.w, "[upgrade/versions] FATAL: Couldn't parse cluster version: %v\n", err)
+		return "", nil, err
+	}
+	return clusterVersionInfo.String(), clusterVersion, nil
```

Comment:
do we need to return an extra string arg? Can't we call `clusterVersion.String()`?